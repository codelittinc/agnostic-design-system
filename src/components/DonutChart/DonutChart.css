@import '../../styles/theme/index.css';

.wrapper {
  position: relative;
}

.tooltip-wrapper {
  background-color: var(--tooltip-background-color, var(--structure-color-white));
  border: var(--tooltip-border, 1px solid #EAECF2);
  border-radius: 8px;
  height: 80px;
  position: absolute;
  width: 80px;
  z-index: 10000;
}

.full-circle-tooltip-wrapper {
  bottom: 170px;
  left: 150px;
}

.tooltip {
  height: inherit;
  width: inherit;
}

.donutchart-full-circle {
  fill: var(--donutchart-full-circle-fill, transparent);
  position: relative;
  stroke: var(--donutchart-full-circle-stroke, var(--structure-color-100));
}

.donutchart-indicator {
  fill: var(--donutchart-indicator-fill, transparent);
  stroke: var(--donutchart-indicator-stroke, #4c7ef6);
  stroke-dasharray: var(--donutchart-indicator-stroke-dasharrray, 0 10000);
  transition: var(--donutchart-indicator-transition, stroke-dasharray 0.3s ease);
}

.donutchart-full-circle:hover,
.donutchart-indicator:hover {
  cursor: pointer;
  stroke-width: 30 !important;
}

.donutchart {
  margin: var(--donutchart-margin, 0 auto);
  border-radius: var(--donutchart-border-radius, 50%);
  display: var(--donutchart-display, block);
  overflow: visible;
}

.donutchart-text {
  align-items: var(--donutchart-text-align-items, center);
  display: var(--donutchart-text-display, flex);
  flex-direction: var(--donutchart-text-flex-direction, column);
  justify-content: var(--donutchart-text-justify-content, center);
  font-family: var(--donutchart-text-font-family, 'Inter');
  fill: var(--donutchart-text-fill, --structure-color-900);
  position: var(--donutchart-text-position, absolute);
  left: var(--donutchart-text-left, 0);
  top: var(--donutchart-text-top, 50%);
  bottom: var(--donutchart-text-bottom, 50%);
  width: var(--donutchart-text-width, 100%);
}

.donutchart-completion-value {
  font-size: var(--donutchart-completion-value-font-size, 22px);
}

.donutchart-message {
  align-items: center;
  color: var(--donutchart-message-color, var(--structure-color-black));
  display: flex;
  font-size: var(--donutchart-message-font-size, 12px);
  font-weight: var(--donutchart-message-font-weight, 400);
  gap: var(--donutchart-message-gap, 5px);
  justify-content: center;
  margin-top: var(--donutchart-message-margin-top, 5px);
}
