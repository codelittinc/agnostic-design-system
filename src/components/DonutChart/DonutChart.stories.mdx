# DonutChart Component

## Type: Simple

## DonutChart Props

|          Prop           |            Type           |                                                                 Description                                                                 | Default | Mandatory |
| :---------------------: | :-----------------------: | :-----------------------------------------------------------------------------------------------------------------------------------------: | :-----: | :-------: |
|        completion       |           number          | Receives the percentage of the filled portion we will show on the chart, where 100 represents the total amount and full width of the chart. |    -    |   true    |
|      iconForMessage     |       React.ReactNode     |                            Receives the icon that would be prepended in the label. You can customize your own icon.                         |    -    |   false   |
|         message         | string \| React.ReactNode |                                         Receives the label that should be displayed below the completion value.                             |    -    |   false   |
|  renderCompletionValue  |           boolean         |            Receives a boolean value that will determine if the completion value should be rendered to the user inside the chart.            |   true  |   false   |
|           size          |           number          |           Specifies the overal size of the Donut Chart. We consider this number to calculate the chart radius and circunference.            |    -    |   true    |
|       strokewidth       |           number          |                                  Specifies the width of the border you want to have on your donut chart.                                    |    -    |   true    |
|    variablesClassName   |           string          |                                         Receives a custom class name for style/css customization.                                           |    -    |   false   |

## Getting Started

```javascript
import React from 'react';
import { DonutChart } from '@codelittinc/agnostic-design-system';

const DonutChart = () => {
  return (
    <DonutChart
      completion={50}
      message="Lorem Ipsum"
      size={200}
      strokewidth={20}
    />
  );
};

export default DonutChart;
```

## CSS Variables Available

|          Property           |    Attribute     | State |
| :-------------------------: | :--------------: | :---: |
|   donutchart-full-circle    |       fill       |       |
|   donutchart-full-circle    |      stroke      |       |
|    donutchart-indicator     |       fill       |       |
|    donutchart-indicator     |      stroke      |       |
|    donutchart-indicator     | stroke-dasharray |       |
|    donutchart-indicator     |    transition    |       |
|         donutchart          |      margin      |       |
|         donutchart          |  border-radius   |       |
|         donutchart          |     display      |       |
|      donutchart-text        |   align-items    |       |
|      donutchart-text        |      bottom      |       |
|      donutchart-text        |     display      |       |
|      donutchart-text        |  flex-direction  |       |
|      donutchart-text        |       fill       |       |
|      donutchart-text        |   font-family    |       |
|      donutchart-text        | justify-content  |       |
|      donutchart-text        |       left       |       |
|      donutchart-text        |     position     |       |
|      donutchart-text        |       top        |       |
|      donutchart-text        |      width       |       |
| donutchart-completion-value |    font-size     |       |
|    donutchart-message       |    font-size     |       |
|    donutchart-message       |   font-weight    |       |
|    donutchart-message       |       gap        |       |
