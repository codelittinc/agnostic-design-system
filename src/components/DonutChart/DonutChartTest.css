.custom-donut-chart {
  --donutchart-full-circle-stroke: var(--secondary-color-100);
  --donutchart-indicator-stroke: var(--secondary-color-400);
  --donutchart-text-font-family: 'Nunito Sans';
  --donutchart-completion-value-font-size: 15px;
  --donutchart-message-color: var(--error-color);
  --donutchart-message-font-size: 20px;
  --donutchart-message-font-weight: 500;
}
