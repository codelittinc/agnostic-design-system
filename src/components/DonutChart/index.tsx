import React, { useState } from 'react';
import styles from '@/components/DonutChart/DonutChart.css';
import classnames from 'classnames';

interface Props {
  completion: number;
  enableTooltip?: boolean;
  fullCircleTooltip?: string | React.ReactNode;
  iconForMessage?: React.ReactNode;
  indicatorTooltip?: string | React.ReactNode;
  message?: string | React.ReactNode;
  renderCompletionValue?: boolean;
  size: number;
  strokewidth: number;
  variablesClassName?: string;
}

const DonutChart = (props: Props) => {
  const {
    completion,
    enableTooltip,
    fullCircleTooltip,
    iconForMessage,
    indicatorTooltip,
    message,
    renderCompletionValue = true,
    size,
    strokewidth,
    variablesClassName
  } = props;

  const [showFullCircleTooltip, setShowFullCircleTooltip] = useState(true);
  const [showIndicatorTooltip, setShowIndicatorTooltip] = useState(false);

  const halfsize = size * 0.5;
  const radius = halfsize - strokewidth * 0.5;
  const circumference = 2 * Math.PI * radius;
  const strokeval = (completion * circumference) / 100;
  const dashval = strokeval + ' ' + circumference;

  const trackstyle = { strokeWidth: strokewidth };
  const indicatorstyle = { strokeWidth: strokewidth, strokeDasharray: dashval };
  const rotateval = 'rotate(-90 ' + halfsize + ',' + halfsize + ')';

  return (
    <div className={classnames(styles.wrapper, variablesClassName)}>
      {showFullCircleTooltip && (
        <div className={classnames(styles['tooltip-wrapper'], styles['full-circle-tooltip-wrapper'])}>
          <div className={styles.tooltip}>{fullCircleTooltip}</div>
        </div>
      )}

      {showIndicatorTooltip && (
        <div className={classnames(styles['tooltip-wrapper'], styles['indicator-tooltip-wrapper'])}>
          <div className={styles.test}>{indicatorTooltip}</div>
        </div>
      )}

      <svg width={size} height={size} className={styles.donutchart}>
        <circle
          r={radius}
          cx={halfsize}
          cy={halfsize}
          transform={rotateval}
          style={trackstyle}
          className={styles['donutchart-full-circle']}
          id="donutchart-full-circle"
          onMouseOver={() => enableTooltip && setShowFullCircleTooltip(true)}
          onMouseLeave={() => enableTooltip && setShowFullCircleTooltip(false)}
        />

        <circle
          r={radius}
          cx={halfsize}
          cy={halfsize}
          transform={rotateval}
          style={indicatorstyle}
          className={styles['donutchart-indicator']}
          onMouseOver={() => enableTooltip && setShowIndicatorTooltip(true)}
          onMouseLeave={() => enableTooltip && setShowIndicatorTooltip(false)}
        />
        <text className={styles['donutchart-text']} x={halfsize} y={halfsize} />
      </svg>

      <span className={styles['donutchart-text']}>
        {renderCompletionValue && (
          <span className={styles['donutchart-completion-value']}>{completion}</span>
        )}

        {message && (
          <span className={styles['donutchart-message']}>
            {iconForMessage} {message}
          </span>
        )}
      </span>
    </div>
  );
};

export default DonutChart;
