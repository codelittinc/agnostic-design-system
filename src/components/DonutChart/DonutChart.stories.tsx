import React from 'react';
import DonutChart from '@/components/DonutChart';
import { Meta } from '@storybook/react';
import mdx from './DonutChart.stories.mdx';
import styles from './DonutChartTest.css';

export default {
  title: 'Components/DonutChart',
  component: DonutChart,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => {
  return <DonutChart {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  completion: 44,
  message: 'No Risk',
  iconForMessage: <span> &#8861; </span>,
  size: 180,
  strokewidth: 10
};

export const WithMessageAsElement = Template.bind({});
WithMessageAsElement.args = {
  completion: 55,
  message: <div>HTML Element Message</div>,
  renderCompletionText: false,
  size: 200,
  strokewidth: 25
};

export const WithHoverTooltip = Template.bind({});
WithHoverTooltip.args = {
  completion: 55,
  enableTooltip: true,
  fullCircleTooltip: 'First Tooltip',
  indicatorTooltip: 'SecondTooltip',
  message: 'Completed',
  renderCompletionValue: false,
  size: 200,
  strokewidth: 25,
}

export const CustomStyles = Template.bind({});
CustomStyles.args = {
  completion: 55,
  message: 'Completed',
  renderCompletionValue: false,
  size: 200,
  strokewidth: 25,
  variablesClassName: styles['custom-donut-chart']
};
