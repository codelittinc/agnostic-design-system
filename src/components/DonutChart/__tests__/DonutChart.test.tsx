import React from 'react';
import { render } from '@testing-library/react';
import DonutChart from '@/components/DonutChart';

describe('DonutChart', () => {
  it('renders correctly', () => {
    const args = {
      completion: 55,
      message: 'Completed',
      size: 150,
      strokewidth: 10
    };
    const { container } = render(<DonutChart {...args} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
