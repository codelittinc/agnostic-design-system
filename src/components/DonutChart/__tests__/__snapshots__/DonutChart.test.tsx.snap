// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutChart renders correctly 1`] = `
<div
  style="position: relative;"
>
  <svg
    class="donutchart"
    height="150"
    width="150"
  >
    <circle
      class="donutchart-full-circle"
      cx="75"
      cy="75"
      r="70"
      style="stroke-width: 10;"
      transform="rotate(-90 75,75)"
    />
    <circle
      class="donutchart-indicator"
      cx="75"
      cy="75"
      r="70"
      style="stroke-width: 10; stroke-dasharray: 241.90263432641405 439.822971502571;"
      transform="rotate(-90 75,75)"
    />
    <text
      class="donutchart-text"
      x="75"
      y="75"
    />
  </svg>
  <span
    class="donutchart-text"
  >
    <span
      class="donutchart-text-val"
    >
      55
    </span>
    <span
      class="donutchart-text-label"
    >
       
      Completed
    </span>
  </span>
</div>
`;
